File: Python/102.py
"""
@no 102
@name Binary Tree Level Order Traversal
"""
class Solution:
    def __appendTreeNodeToListByOrder(self, node, l, order):
        order += 1
        l[order].append(node.val)
        if node.left:
            self.__appendTreeNodeToListByOrder(node.left, l, order)
        if node.right:
            self.__appendTreeNodeToListByOrder(node.right, l, order)

    def __getTreeDepth(self, node):
        if node:
            return max(self.__getTreeDepth(node.left), self.__getTreeDepth(node.right)) + 1
        else:
            return 0

    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if root:
            l = [[] for _ in range(self.__getTreeDepth(root))]
            self.__appendTreeNodeToListByOrder(root, l, -1)
            return l
        else:
            return []

File: JavaScript/102.js
/**
 * @no 102
 * @name Binary Tree Level Order Traversal
 */
let levelOrder = (root) => {
    let ans = [];
    recursion(root, ans, 0);
    return ans;
};

let recursion = (node, ans, depth) => {
    if (node) {
        if (depth === ans.length) ans.push([]);
        ans[depth].push(node.val);
        if (node.left) recursion(node.left, ans, depth + 1);
        if (node.right) recursion(node.right, ans, depth + 1);
    }
};
